GPU monitoring script for Windows version

eissi opened the pull request.

eissi committed:
Create gce-gpu-monitoring-cuda.ps1

eissi committed:
copied from Linux version

eissi committed:
Update README.md

eissi committed:
delete gpuburn as it supports Linux only

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    
+    $query_params = 'gpu_name,gpu_bus_id,'+(($gpu_metrics.Keys|sort-object) -join ",")
+
+    $result = & $nvidia_smi_path --query-gpu=$query_params --format=csv,nounits,noheader
```
If the format is csv, might I suggest us looking into using [`ConvertFrom-CSV` cmdlet](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-csv) to turn the result into an object that's easier to work with?

eissi replied:
Actually, I tried it first, but, it didn't fit to my usage.
The reason follows
1. the header info was not the same with what I have to use. I had to be on par with the column naming of Linux version and the column names of the command don't match to the one of Linux version.
2. The structure I want to form from the CSV isn't actually flat. For example,  the CSV output follows
ColumnA ColumnB ColumnC ColumnD ColumnE
firstname lastname 4               0              100
The structure I want to process later is
@{"firstname,lastname"= @{"columnC"= "4","ColumnD"="0","ColumnE"="100"}}

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    
+}
+
+function get_timeseries_entry{
```
Suggestion: update the function name to `ConvertTo-TimeSeriesEntry` for better aligning with PowerShell best practices.

eissi replied:
I updated the name as advised. Thanks.

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    }
+}
+
+function check_nvidia_smi{
```
Suggestion: update the function name to `Test-NvidiaSMI` for better aligning with PowerShell best practices.

eissi replied:
I updated the name as advised. Thanks.

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    True if the nvidia-smi tool is available.
+    #>
+
+    $installed = test-path -Path $nvidia_smi_path
+    if (-not($installed)){
```
Might consider condense to `if (-Not (Test-Path -Path $nvidia_smi_path)) {`

eissi replied:
I updated the name as advised. Thanks.

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    $result = & $nvidia_smi_path -L 
+    if ($LASTEXITCODE -ne 0){
```
The `if ($?){` should capture the execution success/ failure state more accurately than `$LASTEXITCODE`. I also recommend using `Start-Process` with `-Passthrough` to call the Nvidia SMI binary and determine the state using the returned process obejct.

eissi replied:
I updated the script to use Start-Process.
However, the $? doesn't work for me. $LASTEXITCODE returned the value I want.

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    catch{
+        #if the token is expired, set it again and send the metrics
+        $access_token = Invoke-Expression $access_token_command
+        $headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
+        $headers.Add("Authorization", "Bearer "+$access_token)
+        $headers.Add("Content-Type", 'application/json; charset=utf-8') 
```
These lines may be simplified as following:
```
$headers = @{
  Authorization = "Bearer $access_token"
}
```

And move the `Content-Type` to using the [`Invoke-RestMethod` `-ContentType` parameter](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-restmethod?view=powershell-5.1).

eissi replied:
updated as advised. Thanks.

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+ 
+ Param(
+    #configure the interval, default is 5 and configure bigger than 5 to avoid throttling
+    $interval = 10,
```
It might be a good idea to validate the interval here since anything less than 5 can trigger throttling. An example would be using the [ValidateRange](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/validaterange-attribute-declaration?view=powershell-7.1):

```
[ValidateRange(5..[int]::MaxValue)]
[int]$interval = 10,
```

eissi replied:
good advise. updated as advised. Thanks

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    [switch]$collect_each_core = $false
+)
+
+$nvidia_smi_path = "C:\Program Files\NVIDIA Corporation\NVSMI\nvidia-smi.exe"
```
"C:\Program Files\NVIDIA Corporation\NVSMI\nvidia-smi.exe" can be updated to `"$env:ProgramFiles\NVIDIA Corporation\NVSMI\nvidia-smi.exe"`.

eissi replied:
updated as advaised. Thanks.

LawrenceHwang commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+    $data = @{'timeSeries'=@()}
+
+    #gpu total
+    $now=(get-date).ToString("O")
```
If we intend to post the metric in UTC time, it might be beneficial to ensure we are converting the timestapt to UTC, for example: `(Get-Date).ToUniversalTime().toString("O")`

eissi replied:
Thanks. updated.

LawrenceHwang reviewed (commented):
Hello eissi,

Thanks a bunch for putting this together. While I am not the owner of the repo, I know windows and PowerShell fairly well so shared some thoughts around the code logic and some items (I feel) worth noting. I skipped the code formatting (casing, whitespace...etc) for now as we can address rather easily with proper linting tool.

Please do feel free to ping me if you'd like further clarification around the comments and code. Super happy to help out.


Cheers,
Lawrence

eissi committed:
updated as advised by the reviewer, Lawrence Hwang

eissi committed:
fix 1

eissi committed:
fix 2

eissi committed:
invoke-webrequest update

eissi committed:
formatted

eissi committed:
Merge branch 'main' of https://github.com/eissi/compute-gpu-monitoring

eissi committed:
Code beautifier broke the code

eissi committed:
Don't need -runlevel when register-scheduledtask

eissi committed:
Register-ScheduledTask option changed to run forever

eissi committed:
readme update in testing as the tool is supported in Linux only

eissi committed:
readme update in testing as the tool is supported in Linux only

Just start the monitoring script, start the load generator and visit your Cloud Monitoring metrics explorer to look for metrics like custom.googleapis.com/instance/gpu/utilization.

eissi committed:
Update README.md

eissi committed:
no change. recovered to the original

m-strzelczyk commented on windows/README.md:
```diff
@@ -0,0 +1,85 @@
+## Installation
+
+This instruction assumes installation in `C:\Program Files\NVIDIA Corporation` directory.
+If you change the directory the `nvidia-smi` tool is installed, you have to update your script to be downloaded.
```
```suggestion
If you change the directory the `nvidia-smi` tool is in installed, you have to update your script to use this new directory.
```

m-strzelczyk commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+<#
+This script can be used for monitoring  GPU utilization on Google Compute Engine
+instances with GPUs attached. It requires the nvidia-smi tool to be properly
+installed on the system (https=//developer.nvidia.com/nvidia-system-management-interface).
```
```suggestion
installed on the system (https://developer.nvidia.com/nvidia-system-management-interface).
```

m-strzelczyk commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+installed on the system (https=//developer.nvidia.com/nvidia-system-management-interface).
+
+This script assumes it is being executed on GCE instance and that it has access
+to the metadata server (https=//cloud.google.com/compute/docs/storing-retrieving-metadata).
```
```suggestion
to the metadata server (https://cloud.google.com/compute/docs/storing-retrieving-metadata).
```

m-strzelczyk commented on windows/gce-gpu-monitoring-cuda.ps1:
```diff
@@ -0,0 +1,227 @@
+#>
+ 
+Param(
+    #configure the interval, default is 5 and configure bigger than 5 to avoid throttling
```
Let's make this 10 seconds by default as GCP Monitoring doesn't want more time series anyway. (https://cloud.google.com/monitoring/quotas)

eissi replied:
Sure. no problem and thanks for the detailed review.

m-strzelczyk reviewed (commented):
Hi, thanks a lot for this script! We really missed this kind of feature and now our monitoring solution will be more complete :) 

I added some minor comments. I don't really know PS1 and Windows that well, so I trust that Lawrence with that. I will merge this once I get a chance to run test it out myself and you reply to my suggestions.

Thanks again!

eissi committed:
Update windows/README.md

Co-authored-by: Maciej Strzelczyk <strzelczyk@google.com>

eissi committed:
Update windows/gce-gpu-monitoring-cuda.ps1

Co-authored-by: Maciej Strzelczyk <strzelczyk@google.com>

eissi committed:
Update windows/gce-gpu-monitoring-cuda.ps1

Co-authored-by: Maciej Strzelczyk <strzelczyk@google.com>

eissi committed:
update the minimum interval

eissi commented:
Thanks for reviewing and approving
